jcstress-wrapper/src/main/kotlin/komem/litmus/Codegen.kt

```diff
@@ -0,0 +1,139 @@
+        return "$packagesLine$className.$getter"
+    }
+
+    fun javaThreadFunctionDecl(index: Int) =
+        "private static final Function1<Object, Unit> fT$index = test.getThreadFunctions().get($index);"
+
+    fun javaActorDecl(index: Int) = """
+        @Actor
+        public void t$index() {
+            fT$index.invoke(state);
+        }
+        """.trimIndent()
+
+    fun javaArbiterDecl(): String {
+
+        val jcstressResultClassName = outcomeTypeName + "_Result"
+
+        return if (outcomeVarCount > 1) {
+            """
+@Arbiter
+public void a($jcstressResultClassName r) {
+    List<$outcomeVarType> result = (List<$outcomeVarType>) fA.invoke(state);
+    ${List(outcomeVarCount) { "r.r${it + 1} = result.get($it);" }.joinToString("\n    ")}
+}
+            """.trim()
+        } else {
+            // single values are handled differently
+            """
+            @Arbiter
+            public void a($jcstressResultClassName r) {
+                r.r1 = ($outcomeVarType) fA.invoke(state);
+            }
+            """.trimIndent()
+        }
+    }
+
+    fun jcstressOutcomeDecls(): String {
+        val outcomes = test.outcomeSpec.accepted.associateWith { "ACCEPTABLE" } +
+                test.outcomeSpec.interesting.associateWith { "ACCEPTABLE_INTERESTING" } +
+                test.outcomeSpec.forbidden.associateWith { "FORBIDDEN" }
+
+        // since only AutoOutcome is allowed, each outcome is a list (unless it's a single value)
+        return outcomes.map { (o, t) ->
+            val oId = if (outcomeVarCount > 1) (o as List<*>).joinToString(", ") else o.toString()
+            "@Outcome(id = \"$oId\", expect = $t)"
+        }.joinToString("\n")
+    }
+
+    val jcstressDefaultOutcomeType = when (test.outcomeSpec.default) {
+        LitmusOutcomeType.ACCEPTED -> "ACCEPTABLE"
+        LitmusOutcomeType.FORBIDDEN -> "FORBIDDEN"
+        LitmusOutcomeType.INTERESTING -> "ACCEPTABLE_INTERESTING"
+    }
+
+    return """
+package komem.litmus;
+
+import komem.litmus.testsuite.*;
+
+import kotlin.Unit;
+import kotlin.jvm.functions.Function1;
+
+import org.openjdk.jcstress.annotations.*;
+import org.openjdk.jcstress.infra.results.*;
+
+import static org.openjdk.jcstress.annotations.Expect.*;
+
+import java.util.List;
+
+@JCStressTest
+@State
+${jcstressOutcomeDecls()}
+@Outcome(expect = $jcstressDefaultOutcomeType)
+public class ${test.javaClassName} {
+
+    private static final LitmusTest<Object> test = (LitmusTest<Object>) ${javaTestGetter()};
+    ${List(test.threadCount) { javaThreadFunctionDecl(it) }.joinToString("\n    ")}
+    private static final Function1<Object, Object> fA = test.getOutcomeFinalizer();
+
+    public ${test.javaClassName}() {}
+
+    public Object state = test.getStateProducer().invoke();
+    
+    ${List(test.threadCount) { javaActorDecl(it).padded(4) }.joinToString("\n\n    ")}
+    
+    ${javaArbiterDecl().padded(4)}
+}
+    """.trimIndent()
```

Comment:
Something wrong here with an indentation, no?

Reply:
Also, could you please extract this string (code template) into top-level function (taking all the parameters as arguments)?

Reply:
I believe indentation is ok, top-level function done